sender_test.go

```diff
@@ -2,32 +2,769 @@ package amqp
+	require.NoError(t, session.Close(ctx))
+	cancel()
+	time.Sleep(100 * time.Millisecond)
+	// sending on a closed sender returns ErrLinkClosed
+	if err = snd.Send(context.Background(), NewMessage([]byte("failed"))); !errors.Is(err, ErrSessionClosed) {
+		t.Fatalf("unexpected error %T", err)
+	}
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, client.Close())
+}
+
+func TestSenderSendOnConnClosed(t *testing.T) {
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.PerformOpen("container")
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			return mocks.SenderAttach(0, tt.Name, 0, encoding.ModeUnsettled)
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+	require.NotNil(t, snd)
+	time.Sleep(100 * time.Millisecond)
+
+	require.NoError(t, client.Close())
+	time.Sleep(100 * time.Millisecond)
+	// sending on a closed sender returns ErrLinkClosed
+	if err = snd.Send(context.Background(), NewMessage([]byte("failed"))); !errors.Is(err, ErrConnClosed) {
+		t.Fatalf("unexpected error %T", err)
+	}
+	time.Sleep(100 * time.Millisecond)
+	require.NoError(t, client.Close())
+}
+
+func TestSenderSendOnDetached(t *testing.T) {
+	responder := func(req frames.FrameBody) ([]byte, error) {
+		switch tt := req.(type) {
+		case *mocks.AMQPProto:
+			return []byte{'A', 'M', 'Q', 'P', 0, 1, 0, 0}, nil
+		case *frames.PerformOpen:
+			return mocks.PerformOpen("container")
+		case *frames.PerformBegin:
+			return mocks.PerformBegin(0)
+		case *frames.PerformEnd:
+			return mocks.PerformEnd(0, nil)
+		case *frames.PerformAttach:
+			return mocks.SenderAttach(0, tt.Name, 0, encoding.ModeUnsettled)
+		case *frames.PerformDetach:
+			return mocks.PerformDetach(0, 0, nil)
+		default:
+			return nil, fmt.Errorf("unhandled frame %T", req)
+		}
+	}
+	netConn := mocks.NewNetConn(responder)
+
+	client, err := New(netConn)
+	require.NoError(t, err)
+
+	session, err := client.NewSession()
+	require.NoError(t, err)
+	time.Sleep(100 * time.Millisecond)
+	snd, err := session.NewSender()
+	require.NoError(t, err)
+	require.NotNil(t, snd)
+	time.Sleep(100 * time.Millisecond)
+	// initiate a server-side detach
+	const (
+		errcon  = "detaching"
+		errdesc = "server side detach"
+	)
+	b, err := mocks.PerformDetach(0, 0, &Error{Condition: errcon, Description: errdesc})
+	require.NoError(t, err)
+	netConn.SendFrame(b)
+	time.Sleep(100 * time.Millisecond)
```

Comment:
If I yank out this sleep call what breaks? Would 'Send' just naturally block?

Reply:
Some of these are artifacts from early implementation/experimentation with the mocks package and can probably be removed (they've multiplied via copy/paste).  However, there are some cases where we have to wait until the mux is done, e.g. when checking the link credit after receiving a message, we have to wait until the mux advances (updating credit and other vars along the way) to its "safe point", i.e. paused selecting on channels.
I'll go back through this and remove an unnecessary sleeps.

Reply:
Removed, will fix up other tests in a separate PR.